lib/grandpa/catch-up.go

```diff
@@ -0,0 +1,289 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const catchUpResponseTimeout = time.Second * 5
+
+type catchUp struct {
+	lock sync.Mutex // applied on requestsSent and bestResponse
+
+	requestsSent map[peer.ID]CatchUpRequest
+	bestResponse *atomic.Value // *CatchUpResponse
+
+	catchUpResponseCh chan *CatchUpResponse
+	waitingOnResponse *atomic.Value
+	grandpa           *Service
+}
+
+func newCatchUp(grandpa *Service) *catchUp {
+	return &catchUp{
+		requestsSent:      make(map[peer.ID]CatchUpRequest),
+		grandpa:           grandpa,
+		catchUpResponseCh: make(chan *CatchUpResponse),
+	}
+}
+
+func (c *catchUp) do(to peer.ID, round uint64, setID uint64) error {
+	defer c.waitingOnResponse.Store(false)
+
+	if err := c.sendCatchUpRequest(
+		to, newCatchUpRequest(round, setID),
+	); err != nil {
+		logger.Debugf("failed to send catch up request: %s", err.Error())
+		return err
+	}
+
+	logger.Debugf("successfully sent a catch up request to node %s, for round number %d and set ID %d",
+		to, round, setID)
+
+	c.grandpa.paused.Store(true)
+
+	timer := time.NewTimer(catchUpResponseTimeout)
+	defer timer.Stop()
+
+	select {
+	case <-c.catchUpResponseCh:
+		return nil
+	case <-timer.C:
+		return errors.New("timeout")
+	}
+}
+
+func (c *catchUp) sendCatchUpRequest(to peer.ID, req *CatchUpRequest) error {
+	if c.bestResponse.Load() != nil {
+		logger.Debugf("ignoring neighbour message since we are already processing a catch response")
+		return nil
+	}
+
+	// TODO: Clean up all request sent before 5 min / (neighbour message interval)
+	c.lock.Lock()
+	_, ok := c.requestsSent[to]
+	c.lock.Unlock()
+	if ok {
+		logger.Debugf("ignoring neighbour message since we already sent a catch request to this peer: %s", to)
+		return nil
+	}
+
+	c.waitingOnResponse.Store(true)
+
+	cm, err := req.ToConsensusMessage()
+	if err != nil {
+		return err
+	}
+
+	err = c.grandpa.network.SendMessage(to, cm)
+	if err != nil {
+		return err
+	}
+
+	c.lock.Lock()
+	c.requestsSent[to] = *req
+	c.lock.Unlock()
+	c.grandpa.paused.Store(true)
```

Comment:
The method `do(...)` already call the `c.grandpa.paused.Store(true)`, is it needed to be called here?